connstr/connstr_test.go

```diff
@@ -0,0 +1,268 @@
+			Want: ConnectionParams{
+				CommonParams: CommonParams{
+					Username: "user", Password: Password{"pass"}, ConnectString: "sid",
+					Timezone: time.Local,
+				},
+				ConnParams: ConnParams{
+					ConnClass: "TestClassName", IsSysOper: true,
+				},
+				PoolParams: PoolParams{
+					MinSessions: 3, MaxSessions: 9, SessionIncrement: 3,
+					WaitTimeout: 200 * time.Millisecond, MaxLifeTime: 4000 * time.Second, SessionTimeout: 2000 * time.Second,
+				},
+				onInitStmts: []string{"a", "b"},
+			},
+		},
+	} {
+		tCase := tCase
+		t.Run(tName, func(t *testing.T) {
+			t.Log(tCase.In)
+			P, err := Parse(tCase.In)
+			if err != nil {
+				t.Errorf("%v", err)
+				return
+			}
+			if diff := cmp.Diff(tCase.Want, P, cmpOpts...); diff != "" && tCase.Want.String() != P.String() {
+				t.Errorf("parse of %q\ngot\n\t%#v,\nwanted\n\t%#v\n%s", tCase.In, P, tCase.Want, diff)
+				return
+			}
+			s := setP(P.String(), P.Password.Secret())
+			Q, err := Parse(s)
+			if err != nil {
+				t.Errorf("parseConnString %v", err)
+				return
+			}
+			if diff := cmp.Diff(P, Q, cmpOpts...); diff != "" && P.String() != Q.String() {
+				t.Errorf("params got %+v, wanted %+v\n%s", P, Q, diff)
+				return
+			}
+			if got := setP(Q.String(), Q.Password.Secret()); s != got {
+				t.Errorf("paramString got %q, wanted %q", got, s)
+			}
+		})
+	}
+}
+
+func TestParseTZ(t *testing.T) {
+	for k, v := range map[string]int{
+		"00:00": 0, "+00:00": 0, "-00:00": 0,
+		"01:00": 3600, "+01:00": 3600, "-01:01": -3660,
+		"+02:00": 7200,
+		"+02:03": 7380,
+	} {
+		i, err := ParseTZ(k)
+		t.Logf("Parse(%q): %d %v", k, i, err)
+		if err != nil {
+			t.Fatal(errors.Errorf("%s: %w", k, err))
+		}
+		if i != v {
+			t.Errorf("%s. got %d, wanted %d.", k, i, v)
+		}
+	}
+}
+
+func ExampleAppendLogfmt() {
+	var buf strings.Builder
+	AppendLogfmt(&buf, "user", "scott")
+	AppendLogfmt(&buf, "password", "tiger")
+	AppendLogfmt(&buf, "connectString", "dbhost:1521/orclpdb1?connect_timeout=2")
+	fmt.Println(buf.String())
+	// Output:
+	// user=scott
+	// password=tiger
+	// connectString="dbhost:1521/orclpdb1?connect_timeout=2"
+}
+
+func ExampleConnectString() {
+	var P ConnectionParams
+	P.Username, P.Password = "scott", NewPassword("tiger")
+	P.ConnectString = "dbhost:1521/orclpdb1?connect_timeout=2"
+	P.SessionTimeout = 42 * time.Second
+	P.SetSessionParamOnInit("NLS_NUMERIC_CHARACTERS", ",.")
+	P.SetSessionParamOnInit("NLS_LANGUAGE", "FRENCH")
+	fmt.Println(P.StringWithPassword())
+	// Output:
+	// user=scott password=tiger connectString="dbhost:1521/orclpdb1?connect_timeout=2"
+	// configDir= connectionClass= enableEvents=0 heterogeneousPool=0 libDir= newPassword=
+	// onInit="ALTER SESSION SET NLS_NUMERIC_CHARACTERS = q'(,.)'" onInit="ALTER SESSION SET NLS_LANGUAGE = q'(FRENCH)'"
```

Comment:
For performance, this would ideally execute just one statement with two options like `ALTER SESSION SET NLS_NUMERIC_CHARACTERS=q'(,.)' NLS_LANGUAGE=q'(FRENCH)'`

Reply:
Nice! I didn't know it.